diff --git a/crypto_sign/ml-dsa-44/m4fstack/sign.c b/crypto_sign/ml-dsa-44/m4fstack/sign.c
index a08d6d6..542eabc 100644
--- a/crypto_sign/ml-dsa-44/m4fstack/sign.c
+++ b/crypto_sign/ml-dsa-44/m4fstack/sign.c
@@ -11,6 +11,8 @@
 
 #include "smallntt.h"
 
+#include "pinata_callbacks.h"
+
 /*************************************************
 * Name:        crypto_sign_keypair
 *
@@ -234,7 +236,13 @@ rej:
       poly_small_ntt_copy(scp, tmp0);
       unpack_sk_s1(stmp0, sk, l_idx);
       small_ntt(stmp0->coeffs);
+      if (l_idx == 0) {
+        PINATA_PATCH_mldsa_start_callback();
+      }
       poly_small_basemul_invntt(tmp0, scp, stmp0);
+      if (l_idx == 0) {
+        PINATA_PATCH_mldsa_finish_callback();
+      }
 
       poly_uniform_gamma1_add_stack(tmp0, tmp0, rhoprime, L*(nonce-1) + l_idx, &state.s256);
 
@@ -246,7 +254,6 @@ rej:
       polyz_pack(sig + CTILDEBYTES + l_idx*POLYZ_PACKEDBYTES, tmp0);
   }
 
-
   /* Write signature */
   unsigned int hint_n = 0;
   unsigned int hints_written = 0;
