diff --git a/crypto_sign/ml-dsa-44/m4fstack/sign.c b/crypto_sign/ml-dsa-44/m4fstack/sign.c
index a08d6d6..358344d 100644
--- a/crypto_sign/ml-dsa-44/m4fstack/sign.c
+++ b/crypto_sign/ml-dsa-44/m4fstack/sign.c
@@ -11,6 +11,8 @@
 
 #include "smallntt.h"
 
+#include "pinata_callbacks.h"
+
 /*************************************************
 * Name:        crypto_sign_keypair
 *
@@ -137,6 +139,7 @@ int crypto_sign_signature_ctx(uint8_t *sig,
   uint16_t nonce = 0;
   uint8_t wcomp[K][768];
   uint8_t ccomp[68];
+  uint8_t PINATA_PATCH_rejected_once = 0;
 
   union {
     shake128incctx s128;
@@ -227,6 +230,10 @@ rej:
   poly_challenge_compress(ccomp, tmp0);
   
   /* Compute z, reject if it reveals secret */
+  if (PINATA_PATCH_rejected_once == 0) {
+    PINATA_PATCH_mldsa_start_callback();
+    PINATA_PATCH_rejected_once = 1;
+  }
     for(size_t l_idx=0;l_idx < L; l_idx++){
     if(l_idx != 0){
       poly_challenge_decompress(tmp0, ccomp);
@@ -246,7 +253,6 @@ rej:
       polyz_pack(sig + CTILDEBYTES + l_idx*POLYZ_PACKEDBYTES, tmp0);
   }
 
-
   /* Write signature */
   unsigned int hint_n = 0;
   unsigned int hints_written = 0;
@@ -288,6 +294,7 @@ rej:
     }
     pack_sig_h(sig, tmp0, k_idx, &hints_written);
   }
+  PINATA_PATCH_mldsa_finish_callback();
   pack_sig_h_zero(sig, &hints_written);
   *siglen = CRYPTO_BYTES;
   return 0;
